testing/snapshot.ts

```diff
@@ -33,51 +78,240 @@ function serialize(actual: unknown): string {
 }
 
 /**
- * Write updates to the snapshot file.
+ * Converts a string to a valid JavaScript string which can be wrapped in backticks.
+ *
+ * @example
  *
- * @param context Assert snapshot context
+ * "special characters (\ ` $) will be escaped" -> "special characters (\\ \` \$) will be escaped"
+ */
+function escapeStringForJs(str: string) {
+  return str
+    .replace(/\\/g, "\\\\")
+    .replace(/`/g, "\\`")
+    .replace(/\$/g, "\\$");
+}
+
+let _mode: SnapshotMode;
+/**
+ * Get the snapshot mode.
  */
-function writeSnapshotFileSync(context: AssertSnapshotContext) {
-  const buf = ["export const snapshot = {};"];
-  function escapeStringForJs(str: string) {
-    return str
-      .replace(/\\/g, "\\\\")
-      .replace(/`/g, "\\`")
-      .replace(/\$/g, "\\$");
+function getMode(options: SnapshotOptions) {
+  if (options.mode) {
+    return options.mode;
+  } else if (_mode) {
+    return _mode;
+  } else {
+    return Deno.args.some((arg) => arg === "--update" || arg === "-u")
+      ? "update"
+      : "assert";
   }
-  context.updatedSnapshot.forEach((value, key) => {
-    let formattedValue = escapeStringForJs(value);
-    formattedValue = formattedValue.includes("\n")
-      ? `\n${formattedValue}\n`
-      : formattedValue;
-    const formattedKey = escapeStringForJs(key);
-    buf.push(`\nsnapshot[\`${formattedKey}\`] = \`${formattedValue}\`;`);
-  });
-  const snapshotFilePath = fromFileUrl(context.snapshotFileUrl as URL);
-  ensureFileSync(snapshotFilePath);
-  Deno.writeTextFileSync(snapshotFilePath, buf.join("\n") + "\n");
 }
 
 /**
- * Register a teardown function which writes the snapshot file to disk and logs the number
- * of snapshots updated after all tests have run.
- *
- * This function can safely be called more than once and will only register the teardown
- * function once.
- *
- * @param context Assert snapshot context
+ * Return `true` when snapshot mode is `update`.
  */
-function registerSnapshotTeardown(context: AssertSnapshotContext) {
-  if (context.teardownRegistered) return;
-  globalThis.addEventListener("unload", () => {
-    writeSnapshotFileSync(context);
-    if (context.snapshotUpdatedCount > 0) {
-      console.log(
-        green(bold(`\n > ${context.snapshotUpdatedCount} snapshots updated.`)),
+function getIsUpdate(options: SnapshotOptions) {
+  return getMode(options) === "update";
+}
+
+class AssertSnapshotContext {
+  static contexts = new Map<string, AssertSnapshotContext>();
+
+  /**
+   * Returns an instance of `AssertSnapshotContext`. This will be retrieved from
+   * a cache if an instance was already created for a given snapshot file path.
+   */
+  static fromOptions(
+    testContext: Deno.TestContext,
+    options: SnapshotOptions,
+  ): AssertSnapshotContext {
+    let path: string;
+    if (options.path) {
+      path = resolve(options.path);
```

Comment:
`path` and `dir` are resolved relative to current directory (which is the directory the user invoked `deno test` command) here. I think another option might be to resolve it relative to the test file's directory (`parse(testFilePath).dir`).

I think the latter might be the better choice because the user probably want to place snapshot near the test file. What do you think?

cc @bartlomieju Do you have any preference on this?

Reply:
I agree @kt3k, the latter seems like a better option

Reply:
Whoops :P fixed this